pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,252 @@
+			// return if we fail to delete, but otherwise, keep going
+			if err = g.client.Delete(ctx, deletionSpc); err != nil {
+				if client.IgnoreNotFound(err) != nil {
+					logger.Error(err, fmt.Sprintf("failed to delete SPC %s", spc.Name))
+					return ctrl.Result{}, fmt.Errorf("deleting SPC: %w", err)
+				}
+				return ctrl.Result{}, nil
+			}
+		}
+
+	}
+
+	logger.Info("reconciling Gateway resource with new secret refs for each TLS-enabled listener")
+	if err = util.Upsert(ctx, g.client, gwObj); err != nil {
+		fullErr := fmt.Errorf("failed to reconcile Gateway resource %s: %w", req.Name, err)
+		logger.Error(err, fullErr.Error())
+		g.events.Event(gwObj, corev1.EventTypeWarning, "FailedUpdateOrCreateGateway", fullErr.Error())
+		return ctrl.Result{}, fullErr
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func GenerateGwListenerCertName(gw string, listener gatewayv1.SectionName) string {
+	template := fmt.Sprintf("kv-gw-cert-%s-%s", gw, string(listener))
+	if len(template) > 253 {
+		template = template[:253]
+	}
+
+	return template
+}
+
+func listenerIsKvEnabled(listener gatewayv1.Listener) bool {
+	return listener.TLS != nil && listener.TLS.Options != nil && listener.TLS.Options[tlsCertKvUriAnnotation] != ""
+}
+
+func retrieveClientIdFromListener(ctx context.Context, k8sclient client.Client, namespace string, options map[gatewayv1.AnnotationKey]gatewayv1.AnnotationValue) (string, error) {
+	certUri := string(options[certUriTLSOption])
+	saName := string(options[serviceAccountTLSOption])
+	inputClientId := string(options[clientIdTLSOption])
+
+	// validate user input
+	if certUri != "" && (saName == "" && inputClientId == "") {
+		return "", newUserError(errors.New("user specified cert URI but no serviceaccount or clientid in a listener"), "detected Keyvault Cert URI, but no ServiceAccount or Client ID was provided")
+	}
+	if certUri == "" && (saName != "" || inputClientId != "") {
+		return "", newUserError(errors.New("user specified clientId or SA but no cert URI in a listener"), "detected identity for WorkloadIdentity, but no Keyvault Certificate URI was provided")
+	}
+	if saName != "" && inputClientId != "" {
+		return "", newUserError(errors.New("user specified both serviceaccount and a clientId in the same listener"), "both ServiceAccount name and clientId have been specified, please specify one or the other")
+	}
+
+	// this should never happen since we check for this prior to this function call but just to be safe
+	if certUri == "" && saName == "" && inputClientId == "" {
+		return "", newUserError(errors.New("none of the required TLS options were specified"), "none of cert URI, clientId, or service account were specified")
+	}
+
+	var ret string
+	var err error
+	switch inputClientId {
+	case "":
+		// pull service account
+		wiSa := &corev1.ServiceAccount{}
+		err = k8sclient.Get(ctx, types.NamespacedName{Namespace: namespace, Name: saName}, wiSa)
+		if err != nil {
+			if client.IgnoreNotFound(err) == nil {
```

Comment:
just check if it's the not found error. weird pattern to compare to nil instead of just making the actual comparison you want